### DISCUSSION TITLE
See changed attributes/fields when `trackChanges()` is disabled

### DISCUSSION THREAD

**Discussion by bencroker** [craftcms/cms#13619] [Ideas] [answered]
As a follow-up to https://github.com/craftcms/cms/issues/12967, it appears that when a custom element type has drafts enabled and its `trackChanges()` method returns `false` (the default), the following methods all return an empty array after save.

```php
$element->getDirtyAttributes();
$element->getDirtyFields();
$element->getModifiedFields();
$element->getModifiedFields();
```

This is likely due to this conditional return being executed:
https://github.com/craftcms/cms/blob/793ff3092aaab18e930b7a43ccabef383e821389/src/base/Element.php#L4271-L4273

Is there any way to see what attributes/fields have changed in such a scenario and if not, could this behaviour be tweaked so that there is?

This stems from an issue reported at https://github.com/putyourlightson/craft-blitz/issues/550 in which Blitz is not refreshing the cache when Nodes in the Navigation plugin are modified.

Craft version 4.5.2.

**Comment by brandonkelly** [comment, marked as answer]
`getModifiedFields()` and `getModifiedAttributes()` will only return fields/attributes on a draft which were recorded as modified thanks to `trackChanges()` returning `true` during the draft’s prior saves. So those won’t ever help you if `trackChanges()` is `false`.

Even prior to 4c59bcbb5ea9ed3a6a42ec2be031906abc2475e5 you could check `getDirtyFields()` from `EVENT_AFTER_SAVE` and get the list of fields that were marked as dirty during that request, though.

```php
use craft\base\Element;
use craft\base\ElementInterface;
use craft\base\Event;
use craft\events\ModelEvent;
use craft\helpers\ElementHelper;

Event::on(Element::class, Element::EVENT_BEFORE_SAVE, function(ModelEvent $event) {
    /** @var ElementInterface $element */
    $element = $event->sender;

    if (!ElementHelper::isDraftOrRevision($element)) {
        $dirtyFields = $element->getDirtyFields();
        // ...
    }
});
```

**Reply by bencroker** [reply]
The problem is that `Element::setDirtyAttributes()` is called _only_ when `trackChanges` is `true`. For Node element types in the Navigation, this always evaluates to `false`, so no attributes are ever marked as dirty.
https://github.com/craftcms/cms/blob/b2a1d331943ba2a5530f7c101a5a62a74827999c/src/services/Elements.php#L3343-L3347

Here’s my test code, which always outputs an empty array.

```php
use craft\base\Element;
use craft\base\ElementInterface;
use craft\events\ModelEvent;
use craft\helpers\ElementHelper;
use verbb\navigation\elements\Node;
use yii\base\Event;

Event::on(Node::class, Element::EVENT_BEFORE_SAVE,
    function(ModelEvent $event) {
        /** @var ElementInterface $element */
        $element = $event->sender;

        if (!ElementHelper::isDraftOrRevision($element)) {
            Craft::dd($element->getDirtyFields());
        }
    }
);
```

**Reply by brandonkelly** [reply]
Attributes are a different story because there’s no way for Craft to capture those automatically; each element type would need to construct its dirty attributes manually ([Entry example](https://github.com/craftcms/cms/blob/b2a1d331943ba2a5530f7c101a5a62a74827999c/src/elements/Entry.php#L2000)). So it’s probably just that Node elements aren’t doing that, because they aren’t tracking changes in the first place.

**Reply by bencroker** [reply]
Sorry, I focused on `Element::setDirtyAttributes()` in my previous comment by mistake (Node elements are in fact setting dirty attributes correctly).

It is `$element->getDirtyFields()` that is always ending up empty, although I’m struggling to track down a specific piece of code to show why this might be happening. See my code sample above, which always outputs an empty array.

**Reply by brandonkelly** [reply]
Just got around to testing this and your code is working as expected. Looks like Navigation 2.0.21 was released yesterday with `trackChanges()` now returning `true`: https://github.com/verbb/navigation/commit/fb2e4eee401c8fbae8af2ed76d10058528fbdc34

**Reply by bencroker** [reply]
Yup, with `trackChanges()` returning `true` it works as expected, thanks for looking into it!
